### PR TITLE
Add enigma keymap for contra keyboard

### PR DESCRIPTION
**Pull request by christopherswenson**
This adds a fun custom keymap I made for my contra keyboard, which can emulate an Enigma machine.

This also makes a small change to the contra keyboard `config.h` file, to resolve a redefinition issue with RBG under-glow.

### Enigma Keymap

By default, this layout functions like a normal QWERTY layout. But it also has an in-built Enigma machine emulator. To enable the emulator, press `Fn+.`, and to go back to QWERTY, press `FN+,`.

A demonstration video can be found [here](https://youtu.be/p8kBjP1DCzo).

When Enigma mode is enabled, all letters A-Z will be enciphered using the Enigma emulator. Other keycodes will not be modified. Shift will be sent as normal, so you can choose whether to capitalize the output. If any other modifier (Ctrl, Alt, or Gui) are being pressed, the normal letter will come through.

There are a few key combinations for configuring and diagnosing the Enigma emulator:

* `Fn+Q` resets the emulator to the last-configured settings
* `Fn+W` types out the current Enigma settings, e.g. `B. IA-IIA-IIIA (A, A, A) QW` (meaning reflector B; rotors I, II, and III all at ring setting A; all at position A, and with Q and W swapped on the plugboard)
* `Fn+P` types out 1000 enciphered As, for testing
* `Fn+Backspace` back-rotates the rotors one step, for fixing typos without having to re-type a whole message
* `Fn+Z` followed by A, B, or C is used to set the reflector
* `Fn+X` followed by three characters A-E is used to set the rotors
* `Fn+C` followed by three characters A-Z is used to set the rotor positions
* `Fn+V` followed by three characters A-Z is used to set the ring settings
* `Fn+B` followed by up to 13 pairs of characters A-Z (one at a time, not simulaneously) then the Enter key is used to set the plugboard

### Change to Contra

Contra's `config.h` has this in it:

```
#ifdef RGB_DI_PIN
#define RGBLIGHT_ANIMATIONS
#define RGBLED_NUM 0
#define RGBLIGHT_HUE_STEP 8
#define RGBLIGHT_SAT_STEP 8
#define RGBLIGHT_VAL_STEP 8
#endif
```

And if a keymap attempts to define any of these values differently, compilation fails due to the value being redefined. 

This PR simply removes this block entirely. This is OK because only one other keymap for contra (`maxr1998`) actually defines `RGB_DI_PIN`, and it also attempts to redefine `RGBLIGHT_VAL_STEP` and `RGBLED_NUM`. So this PR actually fixes that keymap, too.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**]
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/contra/keymaps/enigma/keymap.c [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+    */
+    [_RAISE] = LAYOUT_planck_mit(
+        KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_BSPC, 
+        _______, _______, _______, _______, _______, _______, _______, KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC, KC_BSLS, 
+        KC_LSFT, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, 
+        KC_LCTL, _______, KC_LGUI, KC_LALT, _______,      KC_SPC,      _______, KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT
+    ),
+};
+
+void send_character(char letter) {
```
Comment (drashna) [inline keyboards/contra/keymaps/enigma/keymap.c:155]: You could probably use `send_char` for this instead.

File: keyboards/contra/keymaps/enigma/keymap.c [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+              if (setting_plugs) { 
+                  default_settings.plug_count = setting_rotor_n / 2; 
+                  for (int i = 0; i < default_settings.plug_count; i++) {
+                      default_settings.plugs[i * 2] = rotor_settings_progress[i * 2];
+                      default_settings.plugs[i * 2 + 1] = rotor_settings_progress[i * 2 + 1];
+                  }
+                  return save_setting();
+              }
+              break;
+          case EN_A:
```
Comment (drashna) [inline keyboards/contra/keymaps/enigma/keymap.c:489]: This entire block could probably be replaced with just:
```c
    case EN_A ... EN_Z: {
        uint8_t index = keycode - EN_A;
        normal_keycode = KC_A + index;
        letter = 'A' + index;
        break;
        }
```

File: keyboards/contra/keymaps/enigma/keymap.c [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+              send_string("\n");
+              break;
+          case EN_BKSP:
+              rotors_reverse(&current_settings);
+              tap_code(KC_BSPC);
+              break;
+      }
+    }
+    int mods = get_mods();
+    int any_mods = (mods & MOD_MASK_CTRL) || (mods & MOD_MASK_ALT) || (mods & MOD_MASK_GUI);
```
Comment (drashna) [inline keyboards/contra/keymaps/enigma/keymap.c:633]: get_mods is a "uint8_t" type
```suggestion
    uint8_t mods = get_mods();
    bool any_mods = (mods & MOD_MASK_CTRL) || (mods & MOD_MASK_ALT) || (mods & MOD_MASK_GUI);
```

File: keyboards/contra/keymaps/enigma/rules.mk [hunk 4]
```diff
@@ -0,0 +1,4 @@
+AUTO_SHIFT_ENABLE = no
+FORCE_NKRO = yes
```
Comment (drashna) [inline keyboards/contra/keymaps/enigma/rules.mk:2]: This is a define, not a make rule.
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by christopherswenson** [comment]
@drashna Thanks for the great review. I made all the changes you requested. I fixed the repetition of the `case EN_x` stuff a bit differently than you suggested, but basically the same idea.

I also fixed an issue where `Fn+Backspace` was just not mapped to the rotor reverse key code... must have accidentally removed it while I was cleaning up the layouts.

Let me know if you spot anything else that I should change.

**Comment by christopherswenson** [comment]
I'm actually currently working on a few more refactors. Didn't like how big and scary the `process_record_user` function was. I will comment back when done.

**Comment by christopherswenson** [comment]
Okay, I think I'm done refactoring

**Review by keyboard-magpie — APPROVED** [review]

**Comment by keyboard-magpie** [comment]
Exceptionally cool. One to add to the news articles as a "look what QMK is capable of".

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
You are very welcome!
> fixed the repetition of the `case EN_x` stuff a bit differently than you suggested, but basically the same idea.

And yeah, I hate to see repetitive code, especially if it can be condensed down a lot! 

And very cool code!

**Pull request merged**